core/payment/src/processor.rs

```diff
@@ -290,58 +295,91 @@ impl DriverRegistry {
     }
 }
 
-#[derive(Clone)]
+const DB_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
+const REGISTRY_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
+
 pub struct PaymentProcessor {
-    db_executor: DbExecutor,
-    registry: DriverRegistry,
-    in_shutdown: bool,
+    db_executor: Mutex<DbExecutor>,
+    registry: RwLock<DriverRegistry>,
+    in_shutdown: AtomicBool,
 }
 
 impl PaymentProcessor {
     pub fn new(db_executor: DbExecutor) -> Self {
         Self {
-            db_executor,
+            db_executor: Mutex::new(db_executor),
```

Comment:
You assume that `DbExecutor` wasn't cloned, before this function was called.
This mutex protects db from being used in multiple places at the same time by processor, but not in general case